arithmetization/src/main/java/net/consensys/linea/zktracer/module/gas/GasOperation.java

```diff
@@ -15,79 +15,86 @@
 
 package net.consensys.linea.zktracer.module.gas;
 
+import static com.google.common.base.Preconditions.checkArgument;
 import static net.consensys.linea.zktracer.module.constants.GlobalConstants.EVM_INST_LT;
 import static net.consensys.linea.zktracer.module.constants.GlobalConstants.WCP_INST_LEQ;
 import static net.consensys.linea.zktracer.types.Conversions.bigIntegerToBytes;
+import static net.consensys.linea.zktracer.types.Utils.initArray;
 
 import java.math.BigInteger;
 
 import lombok.EqualsAndHashCode;
 import net.consensys.linea.zktracer.container.ModuleOperation;
+import net.consensys.linea.zktracer.module.wcp.Wcp;
 import net.consensys.linea.zktracer.types.UnsignedByte;
 
 @EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = false)
 public class GasOperation extends ModuleOperation {
-  @EqualsAndHashCode.Include GasParameters gasParameters;
-  BigInteger[] wcpArg1Lo;
-  BigInteger[] wcpArg2Lo;
-  UnsignedByte[] wcpInst;
-  boolean[] wcpRes;
-  int CT_MAX;
+  @EqualsAndHashCode.Include GasCall gasCall;
+  private final BigInteger[] wcpArg1Lo;
+  private final BigInteger[] wcpArg2Lo;
+  private final UnsignedByte[] wcpInst;
+  private final boolean[] wcpRes;
+  private final int ctMax;
 
-  public GasOperation(GasParameters gasParameters) {
-    this.gasParameters = gasParameters;
-    CT_MAX = gasParameters.ctMax();
+  public GasOperation(GasCall gasCall, Wcp wcp) {
+    this.gasCall = gasCall;
+    ctMax = compareGasActualAndGasCost() ? 2 : 1;
 
     // init arrays
-    wcpArg1Lo = new BigInteger[CT_MAX + 1];
-    wcpArg2Lo = new BigInteger[CT_MAX + 1];
-    wcpInst = new UnsignedByte[CT_MAX + 1];
-    wcpRes = new boolean[CT_MAX + 1];
+    wcpArg1Lo = initArray(BigInteger.ZERO, ctMax + 1);
+    wcpArg2Lo = initArray(BigInteger.ZERO, ctMax + 1);
+    wcpInst = initArray(UnsignedByte.of(0), ctMax + 1);
+    wcpRes = new boolean[ctMax + 1];
 
     // row 0
     wcpArg1Lo[0] = BigInteger.ZERO;
-    wcpArg2Lo[0] = gasParameters.gasActual();
+    wcpArg2Lo[0] = gasCall.getGasActual();
     wcpInst[0] = UnsignedByte.of(WCP_INST_LEQ);
-    wcpRes[0] = true;
+    final boolean gasActualIsNonNegative = wcp.callLEQ(0, gasCall.getGasActual().longValue());
+    wcpRes[0] = gasActualIsNonNegative; // true
+    checkArgument(gasActualIsNonNegative);
 
     // row 1
     wcpArg1Lo[1] = BigInteger.ZERO;
-    wcpArg2Lo[1] = gasParameters.gasCost();
+    wcpArg2Lo[1] = gasCall.getGasCost();
     wcpInst[1] = UnsignedByte.of(WCP_INST_LEQ);
-    wcpRes[1] = true;
+    final boolean gasCostIsNonNegative = wcp.callLEQ(0, gasCall.getGasCost().longValue());
+    wcpRes[1] = gasCostIsNonNegative; // true
+    checkArgument(gasCostIsNonNegative);
```

Comment:
are the checkARgument really usefull ? It'll break on the constraint check anyway. It's making useless check imho

Reply:
You are right.